### PR TITLE
FIX for #11166 Index Handling Fatal Error

### PR DESCRIPTION
**Pull request by larsroettig**
see #11166
- Add Error Handling
- Implement Test Case

### CODE REVIEW

File: app/code/Magento/Indexer/Model/Indexer.php [hunk 1]
```diff
@@ -418,6 +418,11 @@ public function reindexAll()
                 $state->save();
                 $this->getView()->resume();
                 throw $exception;
+            }catch (\Error $error) {
```
Comment (vrann) [inline app/code/Magento/Indexer/Model/Indexer.php]: - please add space before catch
- is there more specific exception you can catch?

**Review by vrann — CHANGES_REQUESTED** [review]

File: app/code/Magento/Indexer/Model/Indexer.php (hunk 1 above)
Comment (larsroettig) [reply app/code/Magento/Indexer/Model/Indexer.php]: It not so easy to a more specific exception:
Currently,  **php7** throw much different but you must catch all to set the correct state for the indexer. After a fatal, the indexer must be on state **invalid**

php__errors_in_php_7_-_manual

**Review by vrann — APPROVED** [review]

**Comment by hostep** [comment]
Nice, thanks @larsroettig!

I think this fix should be backported to Magento 2.1.x, would be appreciated if someone is willing to do this.

**Comment by larsroettig** [comment]
Hi, @hostep I cannot so easy backport my Fix because 2.1.X allowed to use **php5.6** and this Error handling requires **php7.0** it is broken change for 2.1.X. To catch Engine Error only for one function is in **php5.6** not so simple. 

Magento  2.1.X:
https://github.com/magento/magento2/blob/13d38880e25347d3aec59494813d643fb6975e0d/composer.json#L11

Magento  2.2.X
https://github.com/magento/magento2/blob/4d2609343a950e9ee6b50cc4444469d50d82b139/composer.json#L11

**Comment by hostep** [comment]
@larsroettig: ah yes, you have a valid point there!

We could add a dependency to https://github.com/symfony/polyfill-php70 to get around this, but that's probably a bit out of scope and might introduce new problems.

**Pull request merged**